beamer/agent.py

```diff
@@ -31,107 +37,152 @@ def _get_deployment_block(contract_info: dict[str, ContractInfo]) -> BlockNumber
     return min(request_manager_deployment_block, fill_manager_deployment_block)
 
 
+@dataclass
+class Chain:
+    w3: Web3
+    id: ChainId
+    rpc_url: URL
+
+
+@dataclass
+class Rollup(Chain):
+    contracts_info: dict[str, ContractInfo]
+    contracts: dict[str, Contract]
+    request_manager: RequestManager
+    fill_manager: FillManager
+
+
 class Agent:
     def __init__(self, config: Config):
         self._config = config
         self._stopped = threading.Event()
         self._stopped.set()
         self._init()
 
-    def _init(self) -> None:
-        # Just add one worker, as this effectively serializes the work
-        # This is necessary as we use the account for all resolutions and
-        # would otherwise run into nonce problems
-        self._task_pool = ThreadPoolExecutor(max_workers=1)
-
-        w3_l1 = make_web3(self._config.l1_rpc_url, self._config.account)
-        w3_source = make_web3(self._config.l2a_rpc_url, self._config.account)
-        w3_target = make_web3(self._config.l2b_rpc_url, self._config.account)
-
-        source_chain_id = ChainId(w3_source.eth.chain_id)
-        target_chain_id = ChainId(w3_target.eth.chain_id)
-        chain_ids = {source_chain_id, target_chain_id}
-
-        w3_l1.middleware_onion.add(latest_block_based_cache_middleware)
-
-        source_contracts_info = _get_contracts_info(self._config, w3_source)
-        target_contracts_info = _get_contracts_info(self._config, w3_target)
-
-        source_contracts = make_contracts(w3_source, source_contracts_info)
-        target_contracts = make_contracts(w3_target, target_contracts_info)
-
-        source_request_manager = source_contracts["RequestManager"]
-        source_fill_manager = source_contracts["FillManager"]
-
-        target_fill_manager = target_contracts["FillManager"]
-        target_request_manager = target_contracts["RequestManager"]
-
-        max_validity_period = source_request_manager.functions.MAX_VALIDITY_PERIOD().call()
+    def _init_l1_chain(self) -> Chain:
+        l1_w3 = make_web3(self._config.chains["l1"], self._config.account)
+        l1_w3.middleware_onion.add(latest_block_based_cache_middleware)
+        chain_id = ChainId(l1_w3.eth.chain_id)
+        return Chain(w3=l1_w3, id=chain_id, rpc_url=self._config.chains["l1"])
+
+    def _init_rollups(self) -> dict[ChainId, Rollup]:
+        rollups: dict[ChainId, Rollup] = {}
+        for chain_name, rpc_url in self._config.chains.items():
+            if chain_name == "l1":
+                continue
+            w3 = make_web3(rpc_url, self._config.account)
+            chain_id = ChainId(w3.eth.chain_id)
+            contracts_info = _get_contracts_info(self._config, w3)
+            contracts = make_contracts(w3, contracts_info)
+            request_manager = contracts.get("RequestManager")
+            fill_manager = contracts.get("FillManager")
```

Comment:
Why use `get` here instead of `[]`? We should always have these contracts and, if not, we should crash with an exception.